consensus/spos/bls/subroundEndRound.go

```diff
@@ -551,40 +673,37 @@ func (sr *subroundEndRound) computeAggSigOnValidNodes() ([]byte, []byte, error)
 	return bitmap, sig, nil
 }
 
-func (sr *subroundEndRound) createAndBroadcastHeaderFinalInfo() {
-	leader, errGetLeader := sr.GetLeader()
-	if errGetLeader != nil {
-		log.Debug("createAndBroadcastHeaderFinalInfo.GetLeader", "error", errGetLeader)
-		return
-	}
-
+func (sr *subroundEndRound) createAndBroadcastHeaderFinalInfoForKey(signature []byte, bitmap []byte, leaderSignature []byte, pubKey []byte) bool {
 	cnsMsg := consensus.NewConsensusMessage(
 		sr.GetData(),
 		nil,
 		nil,
 		nil,
-		[]byte(leader),
+		pubKey,
 		nil,
 		int(MtBlockHeaderFinalInfo),
 		sr.RoundHandler().Index(),
 		sr.ChainID(),
-		sr.Header.GetPubKeysBitmap(),
-		sr.Header.GetSignature(),
-		sr.Header.GetLeaderSignature(),
-		sr.GetAssociatedPid([]byte(leader)),
+		bitmap,
+		signature,
+		leaderSignature,
+		sr.GetAssociatedPid(pubKey),
 		nil,
 	)
 
+	// TODO[Sorin next PR]: replace this with the delayed broadcast
 	err := sr.BroadcastMessenger().BroadcastConsensusMessage(cnsMsg)
 	if err != nil {
-		log.Debug("doEndRoundJob.BroadcastConsensusMessage", "error", err.Error())
-		return
+		log.Debug("createAndSendHeaderFinalInfoForKey.BroadcastConsensusMessage", "error", err.Error())
+		return false
 	}
 
 	log.Debug("step 3: block header final info has been sent",
-		"PubKeysBitmap", sr.Header.GetPubKeysBitmap(),
-		"AggregateSignature", sr.Header.GetSignature(),
-		"LeaderSignature", sr.Header.GetLeaderSignature())
+		"PubKeysBitmap", bitmap,
+		"AggregateSignature", signature,
+		"LeaderSignature", leaderSignature)
+
+	return true
 }
 
 func (sr *subroundEndRound) createAndBroadcastInvalidSigners(invalidSigners []byte) {
```

Comment:
the naming might be confusing, as I still kept some of the leader namings.. this is done also by the participants

Reply:
it is called when aggregating signatures, which is done as part of `sendFinalInfo`, which is done from `doEndRoundJobByLeader`, but I've added the following TODO comment: `// TODO[cleanup cns finality]: rename this method, as this will be done by each participant`

Reply:
yes, but at the beginning of this function it checks if it's leader

Reply:
:man_facepalming: true that, good catch, completely missed it

fixed